File: 84-largest-rectangle-in-histogram.java
/*
https://leetcode.com/problems/largest-rectangle-in-histogram/description/

Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.

Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].

The largest rectangle is shown in the shaded area, which has area = 10 unit.

Example:

Input: [2,1,5,6,2,3]
Output: 10
*/

import java.util.*;

class Solution {
    public int largestRectangleArea(int[] hs) {
        if (hs.length == 0) {
            return 0;
        }

        // create a new array
        int[] heights = new int[hs.length + 2];
        for (int i = 0; i < hs.length; i++) {
            heights[i+1] = hs[i];
        }
        heights[0] = 0;
        heights[hs.length+1] = 0;

        Stack<Integer> s = new Stack<>();
        int max = 0;

        for (int i = 0; i < heights.length; i++) {
            // try to push in h
            int h = 0;
            if (i < heights.length) {
                h = heights[i];
            }
            // System.out.printf("current %d: %d\n", i, h);
            
            // pop out all elements that larger than h
            while(!s.empty() && heights[s.peek()] > h) {
                // calc the area based on the poped out height
                int p = s.pop();
                int baseHeight = heights[p];

                int right = i - 1;  // the right most position
                int left = p;
                if (!s.empty()) {
                    left = s.peek() + 1;
                }

                int area = (right - left + 1) * baseHeight;
                if (max < area) {
                    max = area;
                }

                // System.out.printf("  pop %d(%d), [%d - %d], area=%d, max=%d\n", p, baseHeight, left, right, area, max);
            }

            // push in current position
            s.push(i);
        }

        return max;
    }
}


class Main {
    public static void main(String[] args) {
        Solution s = new Solution();
        
        int[] heights = new int[]{2,1,2};
        int area = s.largestRectangleArea(heights);
        System.out.println(area);
    }
}

File: 84-largest-rectangle-in-histogram.py


class Solution:
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        n = len(heights)
        if n <= 0: return 0
        
        leftLower = self.calcLowerArray(heights) 
        rightLower = self.calcLowerArray(list(reversed(heights)))
        # print(leftLower)
        # print(rightLower)
        m = 0
        for i in range(n):
            h = heights[i]
            left = leftLower[i][1]
            right = n - 1 - rightLower[n-i-1][1]
            area = h * (right - left - 1)
            # print(i, h, left, right, area)
            if m < area:
                m = area
        return m

    def calcLowerArray(self, heights):
        stack = MonotonicStack(key = lambda x: x[0])
        lower = []
        for i in range(len(heights)):
            h = heights[i]
            stack.push((h, i))
            if len(stack) <= 1:
                lower.append((0, -1))
            else:
                lower.append(stack[-2])
        return lower

class MonotonicStack:
    def __init__(self, key=None):
        self.stack = []
        if key == None:
            key = lambda x: x
        self.key = key

    def push(self, item):
        out = []
        while len(self.stack) > 0:
            k = self.key(item)
            top = self.key(self.stack[-1])
            if top < k:
                break
            out.append(self.pop())
        self.stack.append(item)
        return out
    
    def pop(self):
        if len(self.stack) == 0:
            return None
        return self.stack.pop()

    def __len__(self):
        return len(self.stack)

    def __getitem__(self, i):
        return self.stack[i]


def main():
    s = Solution()
    h = [2,1,5,6,2,3]
    h = [1, 1]
    print(s.largestRectangleArea(h))

if __name__ == '__main__':
    main()